### PR TITLE
treewide: rewrite autocompletion module and related stuff

### PR DESCRIPTION
**Pull request by diniamo**
## Sanity Checking

[editorconfig]: https://editorconfig.org
[changelog]: https://github.com/NotAShelf/nvf/tree/main/docs/release-notes

- [x] I have updated the [changelog] as per my changes.
- [x] I have tested, and self-reviewed my code.
- Style and consistency
  - [x] I ran **Alejandra** to format my code (`nix fmt`).
  - [x] My code conforms to the [editorconfig] configuration of the project.
  - [x] My changes are consistent with the rest of the codebase.
- If new changes are particularly complex:
  - [x] My code includes comments in particularly complex areas
  - [ ] I have added a section in the manual.
  - [ ] _(For breaking changes)_ I have included a migration guide.
- Package(s) built:
  - [ ] `.#nix` (default package)
  - [ ] `.#maximal`
  - [ ] `.#docs-html`
- Tested on platform(s)
  - [ ] `x86_64-linux`
  - [ ] `aarch64-linux`
  - [ ] `x86_64-darwin`
  - [ ] `aarch64-darwin`

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: modules/plugins/autopairs/nvim-autopairs/config.nix [hunk 1]
```diff
@@ -4,16 +4,15 @@
...
 in {
   config = mkIf cfg.enable {
-    vim.startPlugins = ["nvim-autopairs"];
-
-    vim.pluginRC.autopairs = entryAnywhere ''
-      require("nvim-autopairs").setup({ map_cr = ${boolToString (!config.vim.autocomplete.enable)} })
-    '';
+    vim = {
+      startPlugins = ["nvim-autopairs"];
+      pluginRC.autopairs = entryAnywhere "require('nvim-autopairs').setup(${toLuaObject cfg.setupOpts})";
```
Comment (bloxx12) [inline modules/plugins/autopairs/nvim-autopairs/config.nix:15]:
```suggestion
      pluginRC.autopairs = entryAnywhere ''
      	require('nvim-autopairs').setup(${toLuaObject cfg.setupOpts})
      	'';
```

File: modules/plugins/autopairs/nvim-autopairs/nvim-autopairs.nix [hunk 2]
```diff
@@ -1,21 +1,14 @@
...
-      enable = mkEnableOption "autopairs" // {default = false;};
-
-      type = mkOption {
-        type = enum ["nvim-autopairs"];
-        default = "nvim-autopairs";
-        description = "Set the autopairs type. Options: nvim-autopairs [nvim-autopairs]";
-      };
-    };
+  options.vim.autopairs.nvim-autopairs = {
+    enable = mkEnableOption "autopairs" // {default = false;};
```
Comment (bloxx12) [inline modules/plugins/autopairs/nvim-autopairs/nvim-autopairs.nix:11]:
```suggestion
    enable = mkEnableOption "autopairs";
```

File: modules/plugins/completion/nvim-cmp/nvim-cmp.nix [hunk 3]
```diff
@@ -1,72 +1,103 @@
...
-        next = mkMappingOption "Next item [nvim-cmp]" "<Tab>";
-        previous = mkMappingOption "Previous item [nvim-cmp]" "<S-Tab>";
-        close = mkMappingOption "Close [nvim-cmp]" "<C-e>";
-        scrollDocsUp = mkMappingOption "Scroll docs up [nvim-cmp]" "<C-d>";
-        scrollDocsDown = mkMappingOption "Scroll docs down [nvim-cmp]" "<C-f>";
-      };
+  cfg = config.vim.autocomplete.nvim-cmp;
+in {
+  options.vim.autocomplete.nvim-cmp = {
+    enable = mkEnableOption "nvim-cmp" // {default = false;};
```
Comment (bloxx12) [inline modules/plugins/completion/nvim-cmp/nvim-cmp.nix:17]:
```suggestion
    enable = mkEnableOption "nvim-cmp";
```

File: modules/plugins/notes/obsidian/obsidian.nix [hunk 4]
```diff
@@ -45,10 +45,10 @@ in {
 
         completion = {
           nvim_cmp = mkOption {
-            # if using nvim-cmp, otherwise set to false
+            # If using nvim-cmp, otherwise set to false
             type = bool;
             description = "If using nvim-cmp, otherwise set to false";
-            default = config.vim.autocomplete.type == "nvim-cmp";
+            default = config.vim.autocomplete.nvim-cmp.enable;
```
Comment (NotAShelf) [inline modules/plugins/notes/obsidian/obsidian.nix:51]: This should be an internal option, which is possibly also readOnly. I see no reason a user would want to override this, but if you do see one feel free to make it only internal.

File: modules/plugins/snippets/luasnip/luasnip.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  options.vim.snippets.luasnip = {
+    enable = mkEnableOption "luasnip" // {default = false;};
+    providers = mkOption {
+      type = listOf pluginType;
+      default = ["friendly-snippets"];
+      description = ''
+        The snippet provider packages.
+
+        ::: {.note}
+        These are simply appended to `vim.startPlugins`.
```
Comment (NotAShelf) [inline modules/plugins/snippets/luasnip/luasnip.nix:15]:
```suggestion
        These are simply appended to {option}`vim.startPlugins`.
```

**Review by NotAShelf — CHANGES_REQUESTED** [review]
Added two comments on top of the previous one. Should be good to merge afterwards.

File: modules/plugins/autopairs/nvim-autopairs/config.nix (hunk 1 above)
Comment (diniamo) [reply modules/plugins/autopairs/nvim-autopairs/config.nix:15]: Nonsense

File: modules/plugins/notes/obsidian/obsidian.nix (hunk 4 above)
Comment (diniamo) [reply modules/plugins/notes/obsidian/obsidian.nix:51]: I'm not sure I agree with that, deviating from the raw setupOpts approach should be avoided where possible. Maybe the user wants to disable the completion features, and use the other features of the plugin, and if not, it doesn't matter.

**Review by NotAShelf — APPROVED** [review]

**Comment by C4E5G3** [comment]
I have tried pinning v0.7 but cannot sort out autocomplete, so hopefully when it is all merged, there will be some signposts how to refactor our setup...

**Comment by diniamo** [comment]
There already is, in the release notes.

**Comment by C4E5G3** [comment]
I now get

`The option `vim.autocomplete.sources' does not exist`

**Pull request merged**